Document and test adding `results_metadata` to qp middleware for static viz

markbastian opened the pull request:
Modifying PR to document and use existing fns:
    
The current qp pipeline will correctly propagate custom metadata if a question is derived from a model or if the model is processed like so:
    
    ```clojure
    (qp/process-query
      (assoc-in dataset_query [:info :metadata/dataset-metadata] result_metadata))
    ```
    
This PR fixes rendering unit tests to clarify this and provides a much better working example in the `dev.render-png` ns.

In particular the `dev.render-png` ns has been updated for correctness and tests in `metabase.pulse.render-test` for clarity.

Fixes #36153

markbastian committed:
Adding results_metadata to qp middleware

This adds two middlewares to `metabase.query-processor.middleware.results-metadata`:
- `inject-result-metadata`: Adds `result_metadata` from the context, if present, into the query map during preprocessing.
- `merge-existing-metadata`: A post-processing middleware that merges `results_metadata` (if present) from the query data into the metadata.

This has the effect of preserving existing metadata, which is particularly important for user-curated metadata, such as semantic type overrides.

This may also facilitate addressing the following TODO in the same ns:

```
;; 1. Is there some way we could avoid doing this every single time a Card is ran? Perhaps by passing the current Card
;;    metadata as part of the query context so we can compare for changes
```

markbastian committed:
Preserving result_metadata in query processor

This PR uses `qp.util/combine-metadata` to simplify the combination of provided `:result_metadata` with computed metadata. It also updates pulse rendering code to use this new code path and adds and updates tests to vet this logic.

markbastian committed:
fmt

metamben reviewed (approved):
LGTM

markbastian committed:
Modifying PR to document and use existing fns

The current qp pipeline will correctly propagate custom metadata if a question is derived from a model or if the model is processed like so:

```clojure
(qp/process-query
  (assoc-in dataset_query [:info :metadata/dataset-metadata] result_metadata))
```

This PR fixes rendering unit tests to clarify this and provides a much better working example in the `dev.render-png` ns.

markbastian committed:
Updating test

adam-james-v reviewed (approved):
In this PR's description you mentioned some middleware. 
I don't actually see it in the file changes, and I think that's because you don't need them right?

Instead, what already exists is the ability to do this:  `(assoc-in model-query [:info :metadata/dataset-metadata] model-metadata)`.

Am I understanding that correctly? 
I'm approving this as well because as far as I understand it, it's just a slightly out of date PR description.

markbastian committed:
Adding Pulse w/Metadata Tests

This pr adds a unit test that constructs a dashboard and simulates a pulse email, then checks the resulting HTML for correctness.

A dependency on the [hickory](https://github.com/clj-commons/hickory) library was added to dev. This allows us to easily parse HTML into data and navigate the parse tree.

Note that it does NOT check the attachments for formatting correctness.

Spelling and import consistency changes were also made.
